### PR TITLE
TMA-782: transfer component type process thru lcm

### PR DESCRIPTION
**Pull request by kubamahnert**

### CODE REVIEW

**Comment by kubamahnert** [comment]
https://travis-ci.org/kubamahnert/gooddata-ruby/builds/451500861

File: lib/gooddata/models/project.rb [hunk 1]
```diff
@@ -272,7 +273,8 @@ def transfer_processes(from_project, to_project, options = {})
                          GoodData::Process.deploy_from_appstore(process.path, name: process.name, client: to_project.client, project: to_project)
                        elsif process.component
                          to_process.delete if to_process
-                         GoodData::Process.deploy_component(GoodData::Helpers.symbolize_keys(process.to_hash), project: to_project, client: to_project.client)
+                         process_hash = GoodData::Helpers.symbolize_keys(process.to_hash).deep_merge(additional_hidden_params)
+                         GoodData::Process.deploy_component(process_hash, project: to_project, client: to_project.client)
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: info: Calls 'to_project.client' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/support/comparison_helper.rb [hunk 2]
```diff
@@ -107,6 +107,7 @@ def sanitize_process_for_comparison(process)
         data_dup = process.data.dup
         data_dup['process'].delete('links')
         data_dup['process'].delete('lastDeployed')
+        data_dup['process']['component'].delete('configLocation') if data_dup['process']['component']
```
Comment (yenkins) [inline spec/lcm/integration/support/comparison_helper.rb]: info: Calls 'data_dup['process']' 4 times (DuplicateMethodCall) (reek)

Comment (yenkins) [inline spec/lcm/integration/support/comparison_helper.rb]: info: Calls 'data_dup['process']['component']' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/support/project_helper.rb [hunk 3]
```diff
@@ -87,6 +87,35 @@ def deploy_processes(ads)
...
+
+        component_data = {
+          name: 'lcm-end-to-end-sql-exec',
+          type: :etl,
+          component: {
+            name: 'gdc-etl-sql-executor',
+            version: '1'
+          }
+        }
+        project.deploy_process component_data
```
Comment (yenkins) [inline spec/lcm/integration/support/project_helper.rb]: info: Calls 'project.deploy_process component_data' 2 times (DuplicateMethodCall) (reek)

Comment (yenkins) [inline spec/lcm/integration/support/project_helper.rb]: info: Refers to 'project' more than self (maybe move it to another class?) (FeatureEnvy) (reek)

File: spec/lcm/integration/shared_contexts_for_lcm.rb [hunk 4]
```diff
@@ -170,10 +170,8 @@ def create_workspace_csv(workspaces, client_id_column)
       custom_client_id_column: Support::CUSTOM_CLIENT_ID_COLUMN,
       transfer_all: true,
       conflicting_client_id: conflicting_client_id,
-      # this is very painful, please remove it as soon as possible
-      release_additional_hidden_params: (opts[:release_additional_hidden_params] || {}).to_json,
-      provisioning_additional_hidden_params: (opts[:provisioning_additional_hidden_params] || {}).to_json,
-      rollout_additional_hidden_params: (opts[:rollout_additional_hidden_params] || {}).to_json
+      schedule_additional_hidden_params: (opts[:schedule_additional_hidden_params] || {}).to_json,
+      process_additional_hidden_params: (opts[:process_additional_hidden_params] || {}).to_json,
```
Comment (yenkins) [inline spec/lcm/integration/shared_contexts_for_lcm.rb]: warning: Style/TrailingCommaInHashLiteral: Avoid comma after the last item of a hash. (rubocop)

File: spec/lcm/integration/support/project_helper.rb [hunk 5]
```diff
@@ -87,6 +87,35 @@ def deploy_processes(ads)
...
+                path: 's3://s3_bucket/s3_folder/',
+                accessKey: 's3_access_key',
+                secretKey: 's3_secret_key',
+                serverSideEncryption: true
+              }
+            }
+          }
+        }
+        project.deploy_process component_data
+
```
Comment (yenkins) [inline spec/lcm/integration/support/project_helper.rb]: warning: Layout/EmptyLinesAroundBlockBody: Extra empty line detected at block body end. (rubocop)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/785/ : SUCCESS in 19s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1603/ : FAILURE in 2m 49s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1864/ : SUCCESS in 2m 39s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-build-pipeline/45/ : FAILURE in 4m 07s

File: spec/lcm/integration/support/project_helper.rb [hunk 6]
```diff
@@ -87,6 +87,35 @@ def deploy_processes(ads)
...
+        component_data = {
+          name: 'lcm-end-to-end-csv-download',
+          type: :etl,
+          component: {
+            name: 'gdc-etl-csv-downloader',
+            version: '1',
+            configLocation: {
```
Comment (panjan) [inline spec/lcm/integration/support/project_helper.rb]: `configLocation` is duplicated in lcm_end_to_end_spec.rb

File: lib/gooddata/lcm/actions/synchronize_processes.rb [hunk 7]
```diff
@@ -24,8 +24,8 @@ class SynchronizeProcesses < BaseAction
         description 'Logger'
         param :gdc_logger, instance_of(Type::GdLogger), required: true
 
-        description 'Additional Hidden Parameters'
-        param :additional_hidden_params, instance_of(Type::HashType), required: false
+        description 'Additional Hidden Parameters of Processes'
+        param :process_additional_hidden_params, instance_of(Type::HashType), required: false
```
Comment (panjan) [inline lib/gooddata/lcm/actions/synchronize_processes.rb]: missing unit test for new functionality

Comment (panjan) [reply lib/gooddata/lcm/actions/synchronize_processes.rb]: should we validate nested values of the parameter?

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/795/ : SUCCESS in 6s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1611/ : FAILURE in 2m 54s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1872/ : SUCCESS in 2m 53s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-build-pipeline/53/ : SUCCESS in 8m 26s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/798/ : SUCCESS in 6s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1614/ : FAILURE in 1m 19s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1875/ : SUCCESS in 1m 29s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-build-pipeline/56/ : FAILURE in 7m 20s

File: spec/unit/actions/synchronize_processes_spec.rb [hunk 8]
```diff
@@ -0,0 +1,110 @@
+# encoding: UTF-8
```
Comment (yenkins) [inline spec/unit/actions/synchronize_processes_spec.rb]: warning: Style/Encoding: Unnecessary utf-8 encoding comment. (rubocop)

File: spec/unit/actions/synchronize_processes_spec.rb [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+#
```
Comment (yenkins) [inline spec/unit/actions/synchronize_processes_spec.rb]: warning: Layout/EmptyLineAfterMagicComment: Add an empty line after magic comments. (rubocop)

File: spec/unit/actions/synchronize_processes_spec.rb [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        component: {
+          secretConfig: {
+            key: 'val'
+          }
+        }
+      }
+    }
+  end
+  let(:enriched_process_hash) do
+
```
Comment (yenkins) [inline spec/unit/actions/synchronize_processes_spec.rb]: warning: Layout/EmptyLinesAroundBlockBody: Extra empty line detected at block body beginning. (rubocop)

File: spec/unit/actions/synchronize_processes_spec.rb [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        process: {
+          component: {
+            name: 'etl-csv-uploader',
+            secretConfig: {
+              key: 'val'
+            }
+          }
+        }
+      },
+      {
```
Comment (yenkins) [inline spec/unit/actions/synchronize_processes_spec.rb]: warning: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter. (rubocop)

File: spec/unit/actions/synchronize_processes_spec.rb [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        client: dest_client,
+        project: dest_project
+      })
```
Comment (yenkins) [inline spec/unit/actions/synchronize_processes_spec.rb]: warning: Layout/MultilineMethodCallBraceLayout: Closing method call brace must be on the line after the last argument when opening brace is on a separate line from the first argument. (rubocop)

**Comment by panjan** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/802/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1618/ : FAILURE in 1m 31s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1879/ : FAILURE in 1m 06s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-build-pipeline/58/ : SUCCESS in 8m 02s

**Comment by gdgate** [comment]
Build in pipeline check-extended aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: spec/lcm/integration/support/project_helper.rb [hunk 13]
```diff
@@ -188,9 +216,7 @@ def create_tag_for_fact_n_dataset
 
     def ensure_user(login, domain)
       user = domain.users(login)
-      unless user
-        user = domain.add_user(login: login)
-      end
+      user ||= domain.add_user(login: login)
```
Comment (yenkins) [inline spec/lcm/integration/support/project_helper.rb]: info: Refers to 'domain' more than self (maybe move it to another class?) (FeatureEnvy) (reek)

**Comment by kubamahnert** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/805/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1620/ : SUCCESS in 3m 06s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1881/ : SUCCESS in 2m 59s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-build-pipeline/60/ : SUCCESS in 8m 17s

**Comment by kubamahnert** [comment]
https://travis-ci.org/kubamahnert/gooddata-ruby/builds/454602056

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/806/ : SUCCESS in 6s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1554/ : FAILURE in 11s
- https://checklist.intgdc.com/job/lcm/job/gooddata-ruby-lcm-integration-tests-zuul/114/ : FAILURE in 34m 16s

**Pull request merged**